design/DESIGN.md

```diff
@@ -0,0 +1,99 @@
+
+* Node annotations and labels can become contentious in large clusters.
+* The transport mechanism is not well suited to passing rich data from a node to the controller, which would be needed for features like emitting metrics from the controller that include state information about the fleet.
+* The implementation of the current state machine is difficult to understand, largely due to the cooperative and concurrent design.
+
+While simplifying Brupop’s architecture doesn’t directly resolve many open customer issues, it does provide a path forward for resolving many outstanding issues in a simpler, and likely more robust fashion.
+
+## Solution Overview
+
+Brupop will, by default, continue to rely on Bottlerocket’s client-side update API to determine when to perform an update on any given node — foregoing any complex deployment velocity controls, and instead relying on the wave system built in to Bottlerocket updates. For customers with stateful workloads that require more control, or wish to prevent any changes to their fleet during certain time windows, we will allow customers to disable or enable Brupop via configuration changes. This can later be extended to support more complicated update schedules.
+
+Brupop utilizes Kubernetes’ cordon and [Drain APIs](https://kubernetes.io/docs/tasks/administer-cluster/safely-drain-node/) to prepare a host for updates. Customers with disruption-sensitive workloads should utilize [PodDisruptionBudgets](https://kubernetes.io/docs/concepts/workloads/pods/disruptions/) (which can block `drain` calls that are deemed too disruptive) to ensure that the velocity of updates is not too aggressive for their workload.
+
+Brupop will continue to utilize a DaemonSet of agents on each Bottlerocket nodes. Each of these nodes will also behave as a control loop, operating solely on the target node. Instead of having the independent controller and agent cooperate and pass messages via RPC, we will associate a [Custom Resource](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/) (called BottlerocketNode) with each Bottlerocket node containing status information about the node, as well as a desired state. The `controller` will utilize state information in the `BottlerocketNode` resource to determine what the desired state of the node should be, setting that desired state back into the `spec` of the `BottlerocketNode` resource. The agent on each host will be responsible for updating the `BottlerocketNode` metadata, and driving the local system towards the desired spec set by the controller.
```

Comment:
```suggestion
Brupop will continue to utilize a DaemonSet of agents on each Bottlerocket node. Each of these nodes will also behave as a control loop, operating solely on the target node. Instead of having the independent controller and agent cooperate and pass messages via RPC, we will associate a [Custom Resource](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/) (called BottlerocketNode) with each Bottlerocket node containing status information about the node, as well as a desired state. The `controller` will utilize state information in the `BottlerocketNode` resource to determine what the desired state of the node should be, setting that desired state back into the `spec` of the `BottlerocketNode` resource. The agent on each host will be responsible for updating the `BottlerocketNode` metadata, and driving the local system towards the desired spec set by the controller.
```